core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/chunkedencoding/ChunkedEncodedInputStream.java

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.chunkedencoding;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.http.auth.aws.internal.chunkedencoding.Chunk;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Pair;
+import software.amazon.awssdk.utils.Validate;
+
+
+/**
+ * An implementation of chunk-transfer encoding, but by wrapping an {@link InputStream}.
+ * This implementation supports chunk-headers, chunk-extensions, and trailers.
+ * <p>
+ * Per <a href="https://datatracker.ietf.org/doc/html/rfc7230#section-4.1">RFC-7230</a>, a chunk-transfer encoded message
+ * is defined as:
+ * <pre>
+ *     chunked-body   = *chunk
+ *                      last-chunk
+ *                      trailer-part
+ *                      CRLF
+ *     chunk          = chunk-size [ chunk-ext ] CRLF
+ *                      chunk-data CRLF
+ *     chunk-size     = 1*HEXDIG
+ *     last-chunk     = 1*("0") [ chunk-ext ] CRLF
+ *     chunk-data     = 1*OCTET ; a sequence of chunk-size octets
+ * </pre>
+ */
+@SdkProtectedApi
+public final class ChunkedEncodedInputStream extends InputStream {
+    private static final Logger LOG = Logger.loggerFor(ChunkedEncodedInputStream.class);
+    private static final byte[] CRLF = {'\r', '\n'};
+    private static final byte[] END = {};
+
+    private final InputStream inputStream;
+    private final int chunkSize;
+
+    private final ChunkHeader header;
+    private final List<ChunkExtension> extensions = new ArrayList<>();
```

Comment:
It's weird that we don't actually allow transforming the chunk body. We'll need that for compression.

Should we just combine these into a single transformer?
```java
@FunctionalInterface
interface ChunkTransformer {
    Chunk transform(Chunk chunk);
}

interface Chunk {
    byte[] length();
    List<byte[]> extensions();
    InputStream payload();

    static Chunk create(byte[] length, List<byte[]> extensions, InputStream payload) {
        // ...
    }
}
``` 

I keep `payload()` as an `InputStream` so that we don't have to copy the chunk data into a byte array if we're not going to transform it.

Reply:
I'm not convinced we would use this class for compression, I think it should be used for chunk transfer encoding only. If a request is to be compressed and chunk encoded, it would make sense that it must be compressed first and then chunk encoded. The other way doesn't make sense - if it was encoded and then compressed, you'd have to encode it AGAIN in order to use chunk transfer encoding.

Either way, from what I understand, compressed+chunked can be either `Content-Encoding: gzip, Transfer-Encoding: gzip` or `Transfer-Encoding: gzip, chunked` depending on the intentions, but in both cases, the data should be compressed first and then chunked. To accomplish that, we'd wrap the compressed input stream with the chunked-encoding input stream.

Reply:
That makes sense if the last thing we'd do before the payload is frozen is to load it up into a chunked encoded stream. Unfortunately that's not always the case. There may be things upstream of the caller that has already chunked it, and we have to transform the already-chunked stream.

Reply:
That's a good point, but that might need to be a requirement we have.